### PR TITLE
[T104292598] Refactor the "LRA" training code -> Pytorch Lightning

### PR DESCRIPTION
**Pull request by lisjin**
## What does this PR do?
Refactor LRA [run_tasks.py](https://github.com/facebookresearch/xformers/blob/main/xformers/benchmarks/LRA/run_tasks.py) so that it uses Pytorch Lightning as a trainer.

## Before submitting

- [x] Did you have fun?
  - Make sure you had fun coding 🙃
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/xformers/blob/master/CONTRIBUTING.md)?
- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - [x] N/A
- [ ] Did you make sure to update the docs?
  - [x] N/A
- [ ] Did you write any new necessary tests?
  - [x] N/A
- [ ] Did you update the [changelog](https://github.com/facebookresearch/xformers/blob/master/CHANGELOG.md)? (if needed)
  - [x] N/A

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

**Comment by lisjin** [comment]
@dianaml0 @blefaudeux Hope it's okay to tag you both as reviewers based on the original task (T104292598).

**Comment by blefaudeux** [comment]
> @dianaml0 @blefaudeux Hope it's okay to tag you both as reviewers based on the original task (T104292598).



> @dianaml0 @blefaudeux Hope it's okay to tag you both as reviewers based on the original task (T104292598).

sounds great, thank you for working on this !

File: xformers/benchmarks/LRA/code/model_wrapper.py [hunk 1]
```diff
@@ -134,6 +136,66 @@ def __init__(self, config, model_name):
             * ff_config["hidden_layer_multiplier"]
         )
 
+    def training_step(self, batch, batch_idx):
```
Comment (blefaudeux) [inline xformers/benchmarks/LRA/code/model_wrapper.py:139]: could you type-hint this ? one could expect a tensor here, but it's unwrapped with ** so it's something else (dict I guess), just passing in the expected type helps understanding the convention at a glance I think

File: xformers/benchmarks/LRA/code/model_wrapper.py [hunk 2]
```diff
@@ -182,29 +243,28 @@ def __init__(self, config, model_name):
 
     def forward(self, input_ids_0, input_ids_1, mask_0, mask_1, label):
```
Comment (blefaudeux) [inline xformers/benchmarks/LRA/code/model_wrapper.py:244]: I know that it's not because of this PR, but would it be possible to add some comments here/above/below ? Basically this part of the code base is not of the same quality as the rest I believe (comments/typie hint/code coverage for instance), and I would take the opportunity of this PR to try to raise it little by little ?

File: xformers/benchmarks/LRA/run_tasks.py [hunk 3]
```diff
@@ -77,433 +68,12 @@ def build_model(args: argparse.Namespace, config: Dict) -> nn.Module:
...
         flops = FlopCountAnalysis(model.model, (x, mask, indices))
-        args.logger.info(f"complexity: {round(flops.total()/1e9, 3)} GFlops")
-        args.logger.info(flop_count_str(flops))
+        logging.info(f"complexity: {round(flops.total()/1e9, 3)} GFlops")
+        logging.info(flop_count_str(flops))
 
     return model
 
 
-def build_training_setup(
```
Comment (blefaudeux) [inline xformers/benchmarks/LRA/run_tasks.py:86]: oh wow, this is some sizeable cleanup.. thank you !

File: xformers/benchmarks/LRA/run_tasks.py [hunk 4]
```diff
@@ -563,23 +133,158 @@ def get_arg_parser():
...
+    )
+    logging.warning(
+        f"Requested batch size: {config_training['batch_size']}. Given world\
+                size and grad accumulation, per-gpu batch is\
+                {per_gpu_batch_size}"
+    )
+
+    # Training epochs
+    if accumu_steps > 1:
+        config_training["num_train_steps"] *= accumu_steps
```
Comment (blefaudeux) [inline xformers/benchmarks/LRA/run_tasks.py:193]: I'm not sure that this is still required with lightning, it handles grad accumulation out of the box, right ?

**Comment by blefaudeux** [comment]
Looks good to me @lisjin, thank you for all this work ! Quick braindump
- If possible I would do a second pass, with your critical eye really, trying to comment/simplify/improve the overall code quality here while you're at it (to be clear, I'm really not criticizing this contribution, more the initial quality of this part of the codebase). 
- Would it also be possible to trigger a couple of  jobs to check the results ? [this](https://github.com/facebookresearch/xformers/blob/main/xformers/benchmarks/LRA/batch_submit.py) combined with [that](https://github.com/facebookresearch/xformers/blob/main/xformers/benchmarks/LRA/batch_fetch_results.py) used to be practical to generate the full LRA score matrix in one go on a slurm cluster, it would be great if that still works fine and we can cross check the results ?

File: xformers/benchmarks/LRA/run_tasks.py (hunk 4 above)
Comment (lisjin) [reply xformers/benchmarks/LRA/run_tasks.py:192]: I thought so too but based off [this warning](https://pytorch-lightning.readthedocs.io/en/latest/advanced/training_tricks.html#accumulate-gradients), it sounds like gradient accumulation coupled with DDP behaves differently. I'll look into it in more detail.

Comment (blefaudeux) [reply xformers/benchmarks/LRA/run_tasks.py:192]: this is fine, it just means that prior to the optimizer step the gradients will not be in sync across the fleet, while it's the case without the accumulation. It's a useful warning if you were to peek into the gradients on a per-rank basis, and decide on something from that, in that case triggering the grad acc could mess with your logic. We're not doing that here, simply training over all the ranks, so the default lightning behaviour should be fine

Comment (lisjin) [reply xformers/benchmarks/LRA/run_tasks.py:192]: Ah makes sense—thanks! Will get rid of this block then.

**Comment by lisjin** [comment]
> * Would it also be possible to trigger a couple of  jobs to check the results ? [this](https://github.com/facebookresearch/xformers/blob/main/xformers/benchmarks/LRA/batch_submit.py) combined with [that](https://github.com/facebookresearch/xformers/blob/main/xformers/benchmarks/LRA/batch_fetch_results.py) used to be practical to generate the full LRA score matrix in one go on a slurm cluster, it would be great if that still works fine and we can cross check the results ?

@blefaudeux I ran experiments for `nystrom` attention and the numbers look comparable to the ones reported in [Nystromformer](https://github.com/mlpen/Nystromformer).
| | test_accu_mean |
| --- | --- |
| retrieval | 0.6384 |
| image | 0.3919 |
| pathfinder32-curv_baseline | 0.8597 |
| pathfinder32-curv_contour_length_9 | 0.8018 |
| pathfinder32-curv_contour_length_14 | 0.6592 |
| text | 0.6222 |

File: xformers/factory/model_factory.py [hunk 5]
```diff
@@ -277,7 +277,9 @@ def forward(
 
                 # Apply the optional input masking
                 if encoder_input_mask is not None:
-                    x += encoder_input_mask.unsqueeze(0).unsqueeze(-1)
+                    if x.dim() - encoder_input_mask.dim() > 1:
```
Comment (lisjin) [inline xformers/factory/model_factory.py:280]: I had to add this check to avoid a tensor shape mismatch error.

Comment (blefaudeux) [reply xformers/factory/model_factory.py:280]: oh, well done, thanks for fixing that

**Review by blefaudeux — APPROVED** [review]
LGTM, thanks a million @lisjin ! I'll defer to Diana and Francisco for another validation/landing, but I think that it's a lot cleaner indeed.. thanks for the validation runs also, great PR which was not trivial !

**Review by dianaml0 — APPROVED** [review]
Thanks so much for your contribution! Great to have this improvement and that the LRA results have been validated!

File: xformers/benchmarks/LRA/code/model_wrapper.py [hunk 6]
```diff
@@ -113,11 +117,12 @@ def forward(self, inp_0: torch.Tensor, inp_1: torch.Tensor):
...
 
-class ModelTrunk(nn.Module):
+class ModelTrunk(pl.LightningModule):
     def __init__(self, config, model_name):
         super().__init__()
 
         config_model = config["model"]
+        self.config_training = config["training"]
 
         self.enable_amp = config["training"]["mixed_precision"]
```
[the remark this answers is not in the document]
Comment (lisjin) [reply xformers/benchmarks/LRA/code/model_wrapper.py:127]: A little buried, but it's being used in [`configure_optimizers`](https://github.com/facebookresearch/xformers/blob/5d06c4734fcd4ac35b217dbd7bf2934c1fb87713/xformers/benchmarks/LRA/code/model_wrapper.py#L154).
